EVSL_1.0/SRC/landos.c

```diff
@@ -0,0 +1,220 @@
+#include "blaslapack.h"
+#include "def.h"
+#include "evsl.h"
+#include "internal_proto.h"
+#include "string.h"  //for memset
+#include "struct.h"
+//#include "vect.h"
+
+/**----------------------------------------------------------------------
+ *
+ *    Computes the density of states (DOS, or spectral density)
+ *
+ *    @param[in] *A    matrix A
+ *    @param[in] nvec  number of sample vectors used
+ *    @param[in] msteps number of Lanczos steps
+ *    @param[in] npts number of sample points used ofr the curve
+ *    @param[in] *intv Stores the two intervals of interest \\
+ *      intv[0:1] = [lambda_min, lambda_max]\\
+ *      intv[2:3] = [a b] = interval where DOS is to be computed
+ *
+ *    @param[out] xdos Length-npts long vector, x-coordinate points for
+ *    plotting the DOS. Must be preallocated.
+ *
+ *    @param[out] ydos Length-npts long vector, y-coordinate points for
+ *    plotting the DOS. Must be preallocated.
+ *
+ *    @param[out] neig
+ *
+ *
+ *----------------------------------------------------------------------*/
+
+int LanDos(csrMat *A, const int nvec, int msteps, const int npts, double *xdos,
+           double *ydos, double *neig, const double *const intv) {
+  // Allocations from lanbounds.c
+  double *alp, *bet, nbet, nalp, t, *V;
+  int one = 1;
+  int n;
+
+  n = A->nrows;
+  // Variables that persist through iterations
+  double *v;  // Vector for current iteration
+  Malloc(v, n, double);
+  double *y;  // Stores y values
+  Calloc(y, npts, double);
+  const double lm = intv[0];
+  const double lM = intv[1];
+  const double tolBdwn = 1.e-13 * (abs(lM) + abs(lm));
+  const double aa = max(intv[0], intv[2]);
+  const double bb = min(intv[1], intv[3]);
+  const double kappa = 1.25;
+  const int M = min(msteps, 30);
+  const double H = (lM - lm) / (M - 1);
+  const double sigma = H / sqrt(8 * log(kappa));
+  double sigma2 = 2 * sigma * sigma;
+  // If gaussian small than tol ignore point.
+  const double tol = 1e-08;
+  double width = sigma * sqrt(-2 * log(tol));
+  linspace(aa, bb, npts, xdos);  // xdos = linspace(lm,lM, npts);
+
+  Malloc(alp, msteps, double);
+  Malloc(bet, msteps, double);
+  Malloc(V, (msteps + 1) * n, double);
+
+  for (int m = 0; m < nvec; m++) {
+    randn_double(n, v);  // w = randn(size(A,1),1);
+
+    // Produce predictable vectors for testing
+    // for (int i = 0; i < n; i++) {
+    //   v[i] = 1;
+    // }
+
+    //---------------------------------------
+    // Start of bulk of lanbound.c code
+    //---------------------------------------
+
+    t = DDOT(&n, v, &one, v, &one);
+    t = 1.0 / sqrt(t);  // t is now the number required to normalize vector.
+    DSCAL(&n, &t, v, &one);
+    DCOPY(&n, v, &one, V,
+          &one);  // v = w/norm(w); Might be able to use DNRM2 instead.
+    double wn = 0.0;
+    /*-------------------- main Lanczos loop */
+    int j;
+    for (j = 0; j < msteps; j++) {
+      // w = A*v
+      matvec_genev(A, &V[j * n], &V[(j + 1) * n]);
```

Comment:
matvec_genev was removed in the EVS_Bsol, which I have not yet taken into account. I will have to update my code to accommodate this change.